pkg/kubectl/cmd/config/config.go

```diff
@@ -66,59 +80,323 @@ func NewCmdConfig(f *cmdutil.Factory, out io.Writer) *cobra.Command {
-	filename := ""
+func NewDefaultPathOptions() *PathOptions {
+	ret := &PathOptions{
+		LocalFile:  ".kubeconfig",
+		GlobalFile: os.Getenv("HOME") + "/.kube/.kubeconfig",
+		EnvVar:     clientcmd.RecommendedConfigPathEnvVar,
+		EnvVarFile: os.Getenv(clientcmd.RecommendedConfigPathEnvVar),
+
+		ExplicitFileFlag: clientcmd.RecommendedConfigPathFlag,
+
+		LoadingRules: clientcmd.NewDefaultClientConfigLoadingRules(),
+	}
+	ret.LoadingRules.DoNotResolvePaths = true
+
+	return ret
+}
+
+func (o *PathOptions) Validate() error {
+	if len(o.LoadingRules.ExplicitPath) > 0 {
+		if o.Global {
+			return errors.New("cannot specify both --" + o.ExplicitFileFlag + " and --global")
+		}
+		if o.Local {
+			return errors.New("cannot specify both --" + o.ExplicitFileFlag + " and --local")
+		}
+		if o.UseEnvVar {
+			return errors.New("cannot specify both --" + o.ExplicitFileFlag + " and --envvar")
+		}
+	}
+
+	if o.Global {
+		if o.Local {
+			return errors.New("cannot specify both --global and --local")
+		}
+		if o.UseEnvVar {
+			return errors.New("cannot specify both --global and --envvar")
+		}
+	}
+
+	if o.Local {
+		if o.UseEnvVar {
+			return errors.New("cannot specify both --local and --envvar")
+		}
+	}
+
+	if o.UseEnvVar {
+		if len(o.EnvVarFile) == 0 {
+			return fmt.Errorf("environment variable %v does not have a value", o.EnvVar)
+		}
+
+	}
+
+	return nil
+}
+
+func (o *PathOptions) getStartingConfig() (*clientcmdapi.Config, error) {
+	if err := o.Validate(); err != nil {
+		return nil, err
+	}
+
 	config := clientcmdapi.NewConfig()
 
-	if len(o.specifiedFile) > 0 {
-		filename = o.specifiedFile
-		config = getConfigFromFileOrDie(filename)
+	switch {
+	case o.Global:
+		config = getConfigFromFileOrDie(o.GlobalFile)
+
+	case o.UseEnvVar:
+		config = getConfigFromFileOrDie(o.EnvVarFile)
+
+	case o.Local:
+		config = getConfigFromFileOrDie(o.LocalFile)
+
+		// no specific flag was set, load according to the loading rules
+	default:
+		clientConfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(o.LoadingRules, &clientcmd.ConfigOverrides{})
+		rawConfig, err := clientConfig.RawConfig()
+		if err != nil {
+			return nil, err
+		}
+		config = &rawConfig
+
 	}
 
-	if o.global {
-		if len(filename) > 0 {
-			return nil, "", fmt.Errorf("already loading from %v, cannot specify global as well", filename)
+	return config, nil
+}
+
+// GetDefaultFilename returns the name of the file you should write into (create if necessary), if you're trying to create
+// a new stanza as opposed to updating an existing one.
+func (o *PathOptions) GetDefaultFilename() string {
+	switch {
+	case len(o.LoadingRules.ExplicitPath) > 0:
+		return o.LoadingRules.ExplicitPath
+
+	case o.Global:
+		return o.GlobalFile
+
+	case o.UseEnvVar:
+		return o.EnvVarFile
+
+	case o.Local:
+		return o.LocalFile
+
+		// if there's a local file, use that; else if there's an envvar, use that; else use the global
```

Comment:
Drift.  I'll change that to make it consistent for this pull, but could I convince you to change in the other direction in a separate pull?  --config, local file, env var, $HOME everywhere.

If someone has a local file, they have probably decided that their folder is special in some way.  Everything in the folder pertains to a particular context.  Setting an environment variable is more indicative of, "I want this loaded before (instead of?) my $HOME file".  Thus, changing into a directory that you've decided is special (.kubeconfig isn't created by default), will respect your special settings for this folder.  The environment variable gives you a way to globally indicate that you want something other than $HOME.

/cc @liggitt